File: algorithms/dynamicProgramming/CoinChange.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

// Solution to https://www.hackerrank.com/challenges/coin-change

public class Solution {

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int N = in.nextInt();
        int M = in.nextInt();
        int[] val = new int[M];
        for(int i = 0; i < M; i++){
            val[i] = in.nextInt();
        }

        long[] arr = new long[N+1];
        arr[0] = 1;
        for(int i = 0; i < M; i++){
            for(int j = val[i]; j <= N; j++){
                arr[j] += arr[j - val[i]];
            }
        }
        System.out.println(arr[N]);
    }
}

File: algorithms/dynamicProgramming/coin_change.py
#!/bin/python3

# Solution to https://www.hackerrank.com/challenges/coin-change

import sys

def getWays(n, c):
    a = [0 for x in range(n+1)]
    # one way to make change for 0
    a[0] = 1
    for i in c:
        for j in range(i, n+1):
            a[j] = a[j] + a[j-i]
    return a[n]

n, m = input().strip().split(' ')
n, m = [int(n), int(m)]
c = list(map(int, input().strip().split(' ')))
# Print the number of ways of making change for 'n' units using coins having the values given by 'c'
ways = getWays(n, c)
print(ways)
